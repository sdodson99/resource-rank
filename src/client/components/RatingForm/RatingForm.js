import React, { useEffect, useState } from 'react';
import SelectableRatingStarGroup from '../RatingStars/SelectableRatingStarGroup/SelectableRatingStarGroup';
import PropTypes from 'prop-types';
import styles from './RatingForm.module.css';
import LoadingSpinner from '../LoadingSpinner/LoadingSpinner';

const RatingForm = ({ onSubmit, isSubmitting, error, existingRating }) => {
  const [selectedRating, setSelectedRating] = useState();

  const submitRating = () => onSubmit && onSubmit(selectedRating);

  useEffect(() => {
    setSelectedRating(existingRating);
  }, [existingRating]);

  const ratingChanged = selectedRating !== existingRating;
  const validRating = selectedRating > 0;
  const canSubmitRating = ratingChanged && validRating;

  return (
    <div className={styles.RatingForm} data-testid="RatingForm">
      <SelectableRatingStarGroup
        rating={selectedRating}
        onRatingChanged={setSelectedRating}
        starSize={40}
      />

      <div className="mt-6 flex items-center">
        <button
          className="btn btn-primary"
          onClick={submitRating}
          type="button"
          disabled={!canSubmitRating}
        >
          Submit
        </button>

        {isSubmitting && (
          <div className="ml-4">
            <LoadingSpinner size={25} />
          </div>
        )}
      </div>

      {error && <div className="mt-4 error-text">Failed to submit rating.</div>}
    </div>
  );
};

RatingForm.propTypes = {
  onSubmit: PropTypes.func,
  isSubmitting: PropTypes.bool,
  error: PropTypes.object,
  existingRating: PropTypes.number,
};

RatingForm.defaultProps = {
  isSubmitting: false,
  error: null,
  existingRating: 0,
};

export default RatingForm;
