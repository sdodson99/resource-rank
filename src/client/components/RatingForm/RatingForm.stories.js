import React from 'react';
import RatingForm from './RatingForm';

export default {
  component: RatingForm,
  title: 'Components/RatingForm',
};

const Template = (args) => <RatingForm {...args} />;

export const Default = Template.bind({});
Default.args = {};

export const Submitting = Template.bind({});
Submitting.args = {
  isSubmitting: true,
};

export const Error = Template.bind({});
Error.args = {
  error: new Error(),
};

export const ExistingRating = Template.bind({});
ExistingRating.args = {
  existingRating: 3,
};
