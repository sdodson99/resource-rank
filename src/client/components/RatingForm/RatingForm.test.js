import React, { useEffect, useState } from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { createRenderer } from 'react-test-renderer/shallow';
import RatingForm from './RatingForm';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useEffect: jest.fn(),
  useState: jest.fn(),
}));

describe('<RatingForm />', () => {
  let mockSetSelectedRating;

  beforeEach(() => {
    mockSetSelectedRating = jest.fn();
    useState.mockReturnValue([null, mockSetSelectedRating]);
  });

  afterEach(() => {
    useEffect.mockReset();
    useState.mockReset();
  });

  it('should mount', () => {
    render(<RatingForm />);

    const ratingForm = screen.getByTestId('RatingForm');

    expect(ratingForm).toBeInTheDocument();
  });

  it('should update selected rating on existing rating change', () => {
    const existingRating = 5;
    render(<RatingForm existingRating={existingRating} />);

    useEffect.mock.calls[0][0]();

    expect(mockSetSelectedRating).toBeCalledWith(existingRating);
  });

  it('should submit selected rating on submit', () => {
    const mockOnSubmit = jest.fn();
    const selectedRating = 5;
    useState.mockReturnValue([selectedRating, mockSetSelectedRating]);
    render(<RatingForm onSubmit={mockOnSubmit} />);
    const submitButton = screen.getByText('Submit');

    submitButton.click();

    expect(mockOnSubmit).toBeCalledWith(selectedRating);
  });

  it('should render correctly', () => {
    const component = createRenderer().render(<RatingForm />);

    expect(component).toMatchSnapshot();
  });

  it('should render correctly when submitting', () => {
    const component = createRenderer().render(
      <RatingForm isSubmitting={true} />
    );

    expect(component).toMatchSnapshot();
  });

  it('should render correctly with error', () => {
    const component = createRenderer().render(
      <RatingForm error={new Error()} />
    );

    expect(component).toMatchSnapshot();
  });
});
