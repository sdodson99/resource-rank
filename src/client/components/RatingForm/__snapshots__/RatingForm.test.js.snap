// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RatingForm /> should render correctly 1`] = `
<div
  className="RatingForm"
  data-testid="RatingForm"
>
  <SelectableRatingStarGroup
    maxRating={5}
    onRatingChanged={[MockFunction]}
    rating={null}
    starSize={40}
  />
  <div
    className="mt-6 flex items-center"
  >
    <button
      className="btn btn-primary"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Submit
    </button>
  </div>
</div>
`;

exports[`<RatingForm /> should render correctly when submitting 1`] = `
<div
  className="RatingForm"
  data-testid="RatingForm"
>
  <SelectableRatingStarGroup
    maxRating={5}
    onRatingChanged={[MockFunction]}
    rating={null}
    starSize={40}
  />
  <div
    className="mt-6 flex items-center"
  >
    <button
      className="btn btn-primary"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Submit
    </button>
    <div
      className="ml-4"
    >
      <LoadingSpinner
        color="#292929"
        size={25}
      />
    </div>
  </div>
</div>
`;

exports[`<RatingForm /> should render correctly with error 1`] = `
<div
  className="RatingForm"
  data-testid="RatingForm"
>
  <SelectableRatingStarGroup
    maxRating={5}
    onRatingChanged={[MockFunction]}
    rating={null}
    starSize={40}
  />
  <div
    className="mt-6 flex items-center"
  >
    <button
      className="btn btn-primary"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Submit
    </button>
  </div>
  <div
    className="mt-4 error-text"
  >
    Failed to submit rating.
  </div>
</div>
`;
